### PR TITLE
Generic gear PROTO

### PR DESCRIPTION
**Pull request by ad-daniel**
**Description**
Introduces a generic gear generator PROTO. The tooth profile is inspired on glxgears implementation.

**Images**
model

The boundingObject on the tooth switches between a sphere and a capsule depending on the size of the tip/width of the gear. Here's how they look (in practice they're inside the tooth, but in order to see them I've put them on the edge for the pic)

gear_tip_sph

gear_tip_cyl

### CODE REVIEW

**Review by stefaniapedrazzi — COMMENTED** [review]
Note that the objects documentation is automatically generated by running the `docs/guide/generate_objects_doc.py` script.
We should try to avoid to commit a version that differs from the automatically generated one otherwise we have to manually check the differences every time we regenerate the documentation.

By the way, I noticed some other issues (not related to this PR) in the `generate_objects_doc.py` that produces differences w.r.t the committed files (URLs and derived node) and I will just open an issue about this.
*edit: it has already been fixed on the master branch but not ported to the develop branch. I'm taking care of it.*

**Comment by omichel** [comment]
I would also add a simple example in `projects/samples/howto/gears/` showing at least one motor running one gear that will drive another passive gear (I guess you almost have it already).

**Comment by ad-daniel** [comment]
@stefaniapedrazzi I see, I'll keep it in mind. Do I need to do anything about it this time?

**Comment by stefaniapedrazzi** [comment]
> @stefaniapedrazzi I see, I'll keep it in mind. Do I need to do anything about it this time?

Yes, you need to generate the documentation with the script and commit the auto-generated documentation file instead of your manually written one.
Note that you should first synch with the latest version of the develop branch.

**Comment by ad-daniel** [comment]
Should be good for review now, thanks.

File: docs/guide/samples-howto.md [hunk 1]
```diff
@@ -91,6 +91,16 @@ In the second layout, the four motorized wheels are oriented in the same directi
...
 
+### [gears.wbt]({{ url.github_tree }}/projects/samples/howto/gears/worlds/gears.wbt)
+
+**Keywords**: gears, self-collision
+
+![gears.png](images/samples/gears.thumbnail.jpg) This example shows the transfer of power between two gears through
+physical interaction. The red gear is motorized whereas the blue one is passive, and when the `boundingObject` at the tip of the teeth
+collide the second wheel begins to turn. Using this sort of approach as general purpose transmission is not generally advised but
+does work with some tuning of the `dampingConstant` of the [HingeJoint](../reference/hingejoint.md), the ERP and CFM values in the
+[WorldInfo](../reference/worldinfo.md) node.
```
Comment (stefaniapedrazzi) [inline docs/guide/samples-howto.md:102]: In the documentation we write each sentence on one line and avoid to manually break sentences.
Could you please fix it?

File: projects/samples/howto/gears/worlds/gears.wbt [hunk 2]
```diff
@@ -0,0 +1,65 @@
+#VRML_SIM R2021b utf8
+WorldInfo {
+  basicTimeStep 8
+  coordinateSystem "NUE"
```
Comment (stefaniapedrazzi) [inline projects/samples/howto/gears/worlds/gears.wbt:4]: Could you please also add the title and info?

File: docs/guide/samples-howto.md [hunk 3]
```diff
@@ -91,6 +91,16 @@ In the second layout, the four motorized wheels are oriented in the same directi
 In the third layout, a simple [Ackermann steering geometry](https://en.wikipedia.org/wiki/Ackermann_steering_geometry) is shown.
 Note that more completed Ackermann steering geometry can be achieved using the [`AckermannVehicle` PROTO](../automobile/ackermannvehicle.md), and the [`car` library](../automobile/car-library.md).
 
+### [gears.wbt]({{ url.github_tree }}/projects/samples/howto/gears/worlds/gears.wbt)
+
+**Keywords**: gears, self-collision
+
+![gears.png](images/samples/gears.thumbnail.jpg) This example shows the transfer of power between two gears through
+physical interaction. The red gear is motorized whereas the blue one is passive, and when the `boundingObject` at the tip of the teeth
+collide the second wheel begins to turn. Using this sort of approach as general purpose transmission is not generally advised but
```
Comment (stefaniapedrazzi) [inline docs/guide/samples-howto.md:100]:
```suggestion
collides the second wheel begins to turn. Using this sort of approach as general purpose transmission is not generally advised but
```

File: docs/guide/samples-howto.md [hunk 4]
```diff
@@ -91,6 +91,16 @@ In the second layout, the four motorized wheels are oriented in the same directi
...
+does work with some tuning of the `dampingConstant` of the [HingeJoint](../reference/hingejoint.md), the ERP and CFM values in the
```
Comment (stefaniapedrazzi) [inline docs/guide/samples-howto.md:101]:
```suggestion
does work with some tuning of the `dampingConstant` of the [HingeJoint](../reference/hingejoint.md), and the `ERP` and `CFM` values in the
```

File: projects/objects/gear/protos/Gear.proto [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+  field SFBool          bodyBoundingObject     TRUE       # Enables gear body collision area (cylinder).
+  field SFBool          teethBoundingObject    FALSE      # Enables a boundingObject on each tooth.
+]
+{
+  %{
+    -- load modules
+    local os = require('os')
+    -- parameter retrieval
+    local w = fields.width.value
+    local Z = fields.teeth.value
```
Comment (stefaniapedrazzi) [inline projects/objects/gear/protos/Gear.proto:27]: Wouldn't be better to rename this variable to something more descriptive?

File: projects/objects/gear/protos/Gear.proto [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+    local bodyBoundingObject = fields.bodyBoundingObject.value
+    local teethBoundingObject = fields.teethBoundingObject.value
+    -- parameter sanity check
+    if Z < 8 then
+      io.stderr:write("'teeth' value must bigger than 8. Using default value.\n")
+      Z = fields.teeth.defaultValue
+    end
+
+    if w <= 0.0  then
+      io.stderr:write("'width' value must bigger than zero. Using default value.\n")
```
Comment (stefaniapedrazzi) [inline projects/objects/gear/protos/Gear.proto:41]:
```suggestion
      io.stderr:write("'width' value must be greater than 0. Using default value.\n")
```

File: projects/objects/gear/protos/Gear.proto [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+      w = fields.width.defaultValue
+    end
+
+    if density < 0.0  then
+      io.stderr:write("'density' value must bigger or equal to zero. Using default value.\n")
```
Comment (stefaniapedrazzi) [inline projects/objects/gear/protos/Gear.proto:46]:
```suggestion
      io.stderr:write("'density' value must be greater or equal to 0. Using default value.\n")
```

File: projects/objects/gear/protos/Gear.proto [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+    local Z = fields.teeth.value
+    local innerRadius = fields.innerRadius.value
+    local pitchRadius = fields.pitchRadius.value
+    local toothDepth = fields.toothDepth.value
+    local density = fields.density.value
+    local bodyBoundingObject = fields.bodyBoundingObject.value
+    local teethBoundingObject = fields.teethBoundingObject.value
+    -- parameter sanity check
+    if Z < 8 then
+      io.stderr:write("'teeth' value must bigger than 8. Using default value.\n")
```
Comment (stefaniapedrazzi) [inline projects/objects/gear/protos/Gear.proto:36]:
```suggestion
      io.stderr:write("'teeth' value must be greater than 8. Using default value.\n")
```

File: projects/samples/howto/gears/controllers/gears/Makefile [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+### ---- Webots included libraries ----
+### if you want to use the Webots C API in your C++ controller program:
+### USE_C_API = true
+###
+### ---- Debug mode ----
+### if you want to display the gcc command line for compilation and link, as
+### well as the rm command details used for cleaning:
+### VERBOSE = 1
+###
+###-----------------------------------------------------------------------------
```
Comment (stefaniapedrazzi) [inline projects/samples/howto/gears/controllers/gears/Makefile:66]: For sample controllers we use a shorter Makefile header, you can copy the one from any other howto controller.

**Review by stefaniapedrazzi — CHANGES_REQUESTED** [review]

**Comment by stefaniapedrazzi** [comment]
The Gear PROTO works fine!
There are just some minor issues to fix.

**Review by omichel — CHANGES_REQUESTED** [review]
This [projects/samples/howto/gears/controllers/gears/gears](binary file) should not be committed, but ignored in a local `.gitignore` file.

**Comment by ad-daniel** [comment]
Done

**Review by omichel — APPROVED** [review]
Thank you.

**Review by stefaniapedrazzi — APPROVED** [review]
Thank you

**Comment by stefaniapedrazzi** [comment]
Sorry, I just realized that we should also add an entry in the ChangeLog about this new PROTO.
But this can be done in a separate PR.

**Pull request merged**